Much faster entityNearest using a quadtree

themadcreator opened the pull request:
fixes #3307

themadcreator committed:
Much faster entityNearest using a quadtree

themadcreator commented:
This is SOOO much faster, but there are a couple more things to do

- [ ] Don't rebuild quadtree while panning
- [x] Fix tests (???)

leebyp committed:
Fix tests

leebyp commented:
Note that for the 100k plot, the "rebuilding" of the tree is taking a good 400ms

EDIT: BAM! take out the getter which we don't need anymore because we nuke the cache, and it goes down to 40ms!

leebyp commented on src/plots/plot.ts:
```diff
@@ -388,6 +388,7 @@ export class Plot extends Component {
    * Updates the extents associated with each attribute, then autodomains all scales the Plot uses.
    */
   protected _updateExtents() {
+    this._resetEntityStore();
```
why is this needed here now? isn't all we need done through the `scale.onUpdate`?

themadcreator replied:
Yeah, we can probably remove it but previously we were clearing the cache here and i didn't want to miss it if it was needed. Since "clearing the cache" is just setting to `undefined` it's not gonna be slow

leebyp replied:
we weren't clearing it here originally

leebyp committed:
linting linting

hellochar commented on src/utils/entityStore.ts:
```diff
@@ -3,95 +3,90 @@
+    this._tree = d3.quadtree<T>()
+      // int casting for faster computation. losing sub-pixel precision here is
+      // of no concern.
+      .x((d) => d.position.x | 0)
```
nit: prefer Math.floor() to cast to int. Also do we know that casting to an int actually improves perf significantly?

themadcreator replied:
Alright, i guess math.floor isnt so bad:

![image](https://cloud.githubusercontent.com/assets/569364/24771896/38b114da-1ac4-11e7-82f1-e88737971812.png)

hellochar commented on src/utils/entityStore.ts:
```diff
@@ -3,95 +3,90 @@
-   * the standard Euclidean distance formula.
+   * Looks up the nearest data point using the internal quadtree.
+   *
+   * Note that for performance reasons, the optional filter is **DESTRUCTIVE**
```
So calling `entityNearest` mutates the quadtree? Is the expectation that this is only called once per call to addAll?

themadcreator replied:
It does, which is why i made this comment. There are no guarantees about who will call entityNearest from the outside, but unfortunately creating a copy of the tree every time is quite expensive. What would be ideal is if we could just get the quadtree itself to do the filtering, but that's not how it's implemented

themadcreator replied:
Making a copy means this method takes 10-20msec, whereas without the copy it take <1msec. Thoughts?

hellochar reviewed (approved):
Had some questions but looks good in general

leebyp reviewed (changes requested):
blocking for panning (and zoom?) throttling - unless the removal of the getter makes this fine

themadcreator committed:
Use Math.floor

themadcreator commented:
Remaining tasks are 
- [x] defer resetting entitystore on scale update
- [x] use bounds at entity store build time not nearest query time to avoid destructive update

themadcreator committed:
Debounce entityStore reset.

Adds a debounce method to Utils.Window
Uses the deboucer to debounce `_resetEntityStore`
Also changes entityStore API to pre-filter entities
so that calls to entityNearest are not destructive.

themadcreator committed:
Lint

leebyp commented on src/plots/linePlot.ts:
```diff
@@ -451,7 +451,7 @@ export class Line<X> extends XYPlot<X, number> {
     const chartBounds = this.bounds();
     this.entities().forEach((entity) => {
-      if (!this._entityVisibleOnPlot(entity, chartBounds)) {
+      if (!Utils.Math.within(entity.position, chartBounds)) {
```
oh i guess we can just replace `this._entityVisibleOnPlot` with the math util in `plot.ts` too

themadcreator replied:
yeah i did that which is why i also had to change this here

leebyp commented on src/utils/entityStore.ts:
```diff
@@ -3,95 +3,94 @@
+ *
+ * Note that if the position of your entities changes, you MUST rebuild the
+ * entity store for the `entityNearest` method to work since the quadtree does
+ * not know that its nodes have moved.
```
`and may need to re-filter on the bounds`

themadcreator replied:
yeah but im pretty clear about that in the comment on .addAll and entityNearest
